Add UFOReader and UFOWriter

graphicore opened the pull request.

graphicore committed:
add Bower dependencies

graphicore committed:
Add port of converters module and some tests.

graphicore committed:
Make all Errors subclasses of errors.UFO. Add errors.UFOLib

graphicore committed:
move constants.js directly into ufoLib

graphicore committed:
Add filename constants.

graphicore committed:
More reporting via jshint

graphicore committed:
Add UFOReader and UFOWriter initial Javascript portt

Add base class for UFO[Reader|Writer]

graphicore committed:
UFOReader: fix issues concerning proper error reporting in async execution.

graphicore committed:
Add stub test modules for UFOReader/Writer

Make sure the modules import without errors.

graphicore committed:
debbuging

graphicore committed:
more debugging

graphicore committed:
minor fixes, comments

graphicore committed:
Create a default, empty glyphset on init, if it is missing

graphicore committed:
Fix wrong async argument

graphicore committed:
Make tests pass again

graphicore committed:
Change copyrigh notice in README

Pomax commented on lib/main.js:
```diff
@@ -394,7 +394,7 @@ define(['./errors'], function(errors) {
     function setLike(list, fill /* default true*/) {
         if(fill === undefined) fill = true;
-        var setLike = {};
+        var setLike = Object.create(null);
```
this is considered a JS anti-parttern, with the previous code preferred - why the switch? (always interested in technical details)

graphicore replied:
It is, why?

Pomax replied:
mostly due to the fact that in JS objects (and arrays) are considered primitives, so `a = {}` and `b = []` is clear, whereas `Object.create(null)` is reaching into a part of the API that is unnecessary; in this case, also the use of `null` when JS already "passes in" `undefined` is generally considered making it less JS and more "writing C++/Java with JS syntax" (where you need explicit arguments even if what you need passed in is an undefined/null-reference)

graphicore replied:
> in this case, also the use of null when JS already "passes in" undefined is generally considered making it less JS

sorry, but https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/create

> #### Throws
> Throws a TypeError exception if the proto parameter isn't null or an object.

When I really want only key/value storage I usually use this pattern; also recently more and more often `new Map()`.

I use it when I don't want to create a child of `Object.protototype`, intentionally, to be able to use `key in dict` on objects that I use like maps i.e. only as data. Object is quite overloaded in JS and used for different cases, like class/instances, for inheritance, as the base for a lot of built in types (i.e. function) etc. 

I'd personally would treat this: `obj = {};` as a JavaScript Object and this: `dict = Object.create(null);` as a key/value storage where I know for sure that all keys it has are part of the information it carries.

I'd expect also, if `dict` inherits from another object used as a map object, that all its keys, also the ones inherited, are part of the data, I use this only seldom though.

With a `dict` I can savely use `for(key in dict)` and `if(key in dict)` without checking whether it is something coming from the prototype.
`Object.keys(dict)` would help here, but this and also `Object.hasOwnProperty` etc. are all unnecessary when you start with `Object.create(null)`; In a way these methods are workarounds (for too many uses of js objects) and they require me to write more code.

I'll probably stop this practice in most cases when `for key of mapInstance.keys()` is usable without setting special browser flags or pre-compiling via babel or whatever. However, I'm not convinced by your anti-pattern reasoning. I don't care about "writing C++/Java with JS syntax" arguments, the language is defined in that way.

Pomax replied:
I see, thanks for the detailed explanation. As small bit of info, ES6 introduces a replacement for the older (and pretty shitty) `for ... in ...`, using `for ... of ...` which does object key/val iteration properly, running over own props but not prototype props. This has more immediate applications to arrays, since you can now do `for (let val of arr) { if(...) break; ... }` and not worry about `v` becoming prototype keys like `map` or `length`, but it works the same for objects, and does pair extraction too, using `for (let[key,value] of obj) { ... }`

graphicore replied:
I never tried ufoJS on IE ... I'd already be happy if cutting edge IE supports it. Also, NodeJS is an environment for ufoJS, though since io.js is part of it, it makes pretty good progress. So, probably `for ... of ` is also there.

davelab6 replied:
On 19 April 2016 at 13:47, Lasse Fister <notifications@github.com> wrote:

> cutting edge IE


Very good! ;)

graphicore replied:
On 19.04.2016 20:16, Dave Crossland wrote:
>
> In lib/main.js
> <https://github.com/graphicore/ufoJS/pull/66#discussion_r60282664>:
>
> > @@ -394,7 +394,7 @@ define(['./errors'], function(errors) {
> >      // for easy membership testing with the "in" keyword
> >      function setLike(list, fill /* default true*/) {
> >          if(fill === undefined) fill = true;
> > -        var setLike = {};
> > +        var setLike = Object.create(null);
> On 19 April 2016 at 13:47, Lasse Fister ***@***.***> wrote: cutting
> edge IE
> Very good! ;)
>
Yeah, it's subtle.

Pomax commented on lib/plistLib/main.js:
```diff
@@ -127,7 +127,8 @@ define([
         getType: function(value) {
-            if(value instanceof IntObject)
+            if(value instanceof IntObject
+                    || typeof value === 'number' && value === (value|0))
```
this looks like it's missing parentheses, unless you intended both `typeof value === 'number' && value === (value|0)` and `value instanceof IntObject && value === (value|0)` to be necessary. Otherwise this probably wants to be
```
if(value instanceof IntObject
    || (typeof value === 'number' && value === (value|0)))
```

graphicore replied:
I don't get it:

```
> true || false && false
true
```

Pomax replied:
That relies on the engine shortcutting the OR. It probably will, but there's nothing in the spec that says it has to, so the extra parens harder your code against quirks in other engines.

graphicore replied:
I'm pretty sure this is the only way to interpret it. How can this not be in the spec?

Pomax replied:
Ah, no I just reread the spec and it is in there so I guess it's safe. One neat thing I did find (always new things to find in this spec) is that "The value produced by a && or || operator is not necessarily of type Boolean. The value produced will always be the value of one of the two operand expressions." which isn't a problem here, but not something you'd expect, either.

graphicore replied:
> The value produced will always be the value of one of the two operand expressions."

I use that a lot, but that may be a bit bad for beginners to read, i.e.:
```
var myVal = input || 15;
```
to set a default etc.

It is maybe a bad habit, as I said, for beginners its cryptic. But then again, I'd look it up if I don't know what it does and then learn something more about the language. So a beginner with good manual-reading habits would learn.


Also, it's quite useful to keep a javascript console open for such cases. And, I want to point out that I recently created a bug with something similar (fixed it again). So yeah, even if you are used to it, it still gets you sometimes: https://github.com/graphicore/Atem-CPS/commit/3b8109ab1b529c7057880e265d15bda1dbebcf18

Pomax commented on lib/plistLib/main.js:
```diff
@@ -250,8 +252,8 @@ define([
-                for(key in dict) { 
-                    if(!dict.hasOwnProperty(key) || dict[key] === undefined || dict[key] === null)
+                for(key in dict) {
+                    if(!Object.prototype.hasOwnProperty.call(dict, key) || dict[key] === undefined || dict[key] === null)
```
again quite interested in why this was changed, as the former is the "correct" JS pattern. What made it not work?

graphicore replied:
If you do `Object.create(null)` you don't have `Object` as prototype and thus not `hasOwnProperty` as a method.

Pomax replied:
Interesting. Trying to understand what that's a good thing, if you start with a {} primitive, you keep the hasOwnProperty, without any performance penalties (prototypes are singletons), what made you decide you needed a prototype-detached object?

graphicore replied:
I described above my reasoning for creating objects as key/value storage like `Object.create(null)` In this case I allowed such Objects to be passed as arguments to plist serialization. This implementation is probably not very smart and just a reaction to a bug I had (fixed it though).

Better would be to check whether `Object.prototype` is the beginning of the prototype chain and then use the old code or if `null` is the beginning of the prototype chain and then just use `for key in`. This would work with what I wrote above, how I expect a "dict" to behave, i.e. that its prototypes are also part of the data.

graphicore committed:
Differentiate between map-like objects 'Object.create(null)' and normal '{}' JS-Objects.

See: https://github.com/graphicore/ufoJS/pull/66#discussion-diff-60263222
And: https://github.com/graphicore/ufoJS/pull/66#discussion-diff-60262715
Thanks @Pomax

graphicore committed:
make jshint happy

graphicore committed:
Merge branch 'ReaderWriter' of github.com:graphicore/ufoJS into ReaderWriter

graphicore committed:
Bugfixes for UFOReader

graphicore committed:
FIX: condition needed braces.
